subsys/storage/storage_area/storage_area_store.c

```diff
@@ -0,0 +1,1097 @@
+
+	if ((cb == NULL) || (cb->move == NULL)) {
+		return 0;
+	}
+
+	const size_t erase_size = store->area->erase_size;
+	const size_t sec_size = store->sector_size;
+	const size_t dsector = data->sector;
+	const size_t dloc = data->loc;
+	const size_t dwrapcnt = data->wrapcnt;
+	int rc = 0;
+
+	for (size_t loop = 0; loop < 2; loop++) {
+		size_t rscnt = 0U;
+
+		while (((data->sector * sec_size) % erase_size) != 0U) {
+			store_reverse(store);
+			rscnt++;
+		}
+
+		store_reverse(store);
+		rscnt++;
+		if (loop != 0U) {
+			rc = store_advance_pcb(store, cb);
+			break;
+		}
+
+		struct storage_area_record walk = {
+			.store = (struct storage_area_store *)store,
+		};
+		size_t mrcnt = 0U; /* cnt records that should be moved */
+		size_t vrcnt = 0U; /* cnt records that are moved and valid */
+
+		walk.sector = data->sector;
+		sector_advance(store, &walk.sector, store->spare_sectors + 1U);
+		for (size_t cnt = 0U; cnt < erase_size / sec_size; cnt++) {
+			walk.loc = 0U;
+			walk.size = 0U;
+			while (store_record_next_in_sector(&walk, true, true) ==
+			       0) {
+				if ((cb->move(&walk)) &&
+				    (store_record_valid(&walk))) {
+					mrcnt++;
+				}
+			}
+
+			sector_advance(store, &walk.sector, 1U);
+		}
+
+		data->sector = dsector;
+		data->loc = dloc;
+		data->wrapcnt = dwrapcnt;
+
+		if (mrcnt == 0U) {
+			break;
+		}
+
+		walk.sector = data->sector;
+		while (((walk.sector * sec_size) % erase_size) != 0U) {
+			sector_reverse(store, &walk.sector, 1U);
+		}
+
+		for (size_t cnt = 0U; cnt < rscnt; cnt++) {
+			walk.loc = 0U;
+			walk.size = 0U;
+			/* walk without recovery */
+			while (store_record_next_in_sector(&walk, true, false) ==
+			       0) {
+				if (store_record_valid(&walk)) {
+					vrcnt++;
+				}
+			}
+
+			sector_advance(store, &walk.sector, 1U);
+		}
+
+		if (vrcnt >= mrcnt) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+static size_t store_iovec_size(const struct storage_area_iovec *iovec,
+			       size_t iovcnt)
+{
+	size_t rv = 0U;
+
+	for (size_t i = 0U; i < iovcnt; i++) {
+		rv += iovec[i].len;
+	}
+
+	return rv;
+}
+
+static int store_writev(const struct storage_area_store *store,
+			const struct storage_area_iovec *iovec, size_t iovcnt)
+{
+	struct storage_area_store_data *data = store->data;
+	const size_t len =
+		SAS_HDRSIZE + store_iovec_size(iovec, iovcnt) + SAS_CRCSIZE;
+
+	if ((store->sector_size - len) < data->loc) {
+		return -ENOSPC;
+	}
+
+	const struct storage_area *area = store->area;
+	const size_t secpos = data->sector * store->sector_size;
+	struct storage_area_iovec wr[iovcnt + 2];
+	uint8_t hbuf[SAS_HDRSIZE];
+	uint8_t cbuf[SAS_CRCSIZE + SAS_ALIGNUP(len, area->write_size) - len];
+	uint32_t crc = SAS_CRCINIT;
+	size_t crc_skip = store->crc_skip;
+	int rc;
+
+	hbuf[0] = SAS_MAGIC;
+	hbuf[1] = data->wrapcnt;
+	sys_put_le16((uint16_t)store_iovec_size(iovec, iovcnt), &hbuf[2]);
+	wr[0].data = hbuf;
+	wr[0].len = sizeof(hbuf);
+	wr[iovcnt + 1].data = cbuf;
+	wr[iovcnt + 1].len = sizeof(cbuf);
```

Comment:
The header is used only in 2 places in the code, it is respectively packed or unpacked in these routines.

I don't see the benefit of a struct. Whatever struct is used would need to be converted into a `uint8_t header[4]` because this ensures how it is written regardless of endianness.

Reply:
Renamed the used buffer to `header[]`, this can now be navigated easily.

Reply:
Every time something will need to change in header, you have literally go through code and repack header.
Second issue: structure are documentation, you are not documenting structure of the storage.
Third issue: as far as I understand callback get `storage_area_record` which points to header, and they have to skip that, so callback actually have a header but have to manually parse it.